pageserver/src/tenant/storage_layer/layer_contents/virtual_value.rs

```diff
@@ -0,0 +1,207 @@
+                        let new_vv = match virtual_value {
+                            VirtualValue::NaturalImage(img) => VirtualValue::ClosedLineage {
+                                image: img,
+                                lsns: vec![new_lsn],
+                                records: vec![new_rec],
+                            },
+                            VirtualValue::NaturalWalRecord(vv_start) => {
+                                if vv_start.will_init() {
+                                    VirtualValue::ClosedRecLineage {
+                                        image_rec: vv_start,
+                                        lsns: vec![new_lsn],
+                                        records: vec![new_rec],
+                                    }
+                                } else {
+                                    VirtualValue::OpenLineage {
+                                        lsns: vec![new_lsn],
+                                        records: vec![vv_start, new_rec],
+                                    }
+                                }
+                            }
+                            VirtualValue::ClosedLineage {
+                                image,
+                                mut lsns,
+                                mut records,
+                            } => {
+                                lsns.push(new_lsn);
+                                records.push(new_rec);
+
+                                VirtualValue::ClosedLineage {
+                                    image,
+                                    lsns,
+                                    records,
+                                }
+                            }
+                            VirtualValue::ClosedRecLineage {
+                                image_rec,
+                                mut lsns,
+                                mut records,
+                            } => {
+                                lsns.push(new_lsn);
+                                records.push(new_rec);
+
+                                VirtualValue::ClosedRecLineage {
+                                    image_rec,
+                                    lsns,
+                                    records,
+                                }
+                            }
+                            VirtualValue::OpenLineage {
+                                mut lsns,
+                                mut records,
+                            } => {
+                                lsns.push(new_lsn);
+                                records.push(new_rec);
+
+                                VirtualValue::OpenLineage { lsns, records }
+                            }
+                        };
+
+                        self.state = Some((start_lsn, new_vv));
+
+                        None
+                    }
+                }
+            }
+        }
+    }
+
+    pub fn finish(mut self) -> Option<(Lsn, VirtualValue)> {
```

Comment:
If not doing the assert at drop, could add a `#[must_use]` here to nag on "cancellation through not handling the resulting value":

```rust
vvb.finish();
```

```suggestion
    #[must_use]
    pub fn finish(mut self) -> Option<(Lsn, VirtualValue)> {
```

Reply:
If this route is selected, I think the `fn push` should have a `#[must_use]`.